### ISSUE TITLE
As a system integrator, I would appreciate a JSON Schema for validating my dataset JSON before uploading via API

### ISSUE THREAD

**Issue by poikilotherm** [IQSS/dataverse#7173]
This is related to AUSSDA/pyDataverse#48 and to `dvcli` as a CLI tool for Dataverse.

When creating a new dataset via the web UI, you will be provided with a nice interface and validation before a dataset is created. What is required, what is available as a field, etc is nicely integrated into the UI both for users and curators.

However, this is not the case with uploading new datasets via API. Before you send a JSON representation, there is not possibility to validate the dataset in terms of metadata schemas, required fields, etc.

It would be nice to provide an API endpoint to retrieve a [JSON Schema](https://json-schema.org/) for a given Dataverse, that contains precise constraints and requirements, what you dataset JSON has to look like.

This is useful not only for pre-creation validation, but also for automatic creation of options on command lines (think autocompletion, ncurses interfaces, ...) or [client-side forms](https://json-schema.org/implementations.html#web-ui-generation).

**Comment by poikilotherm** [comment]
The all mighty @pdurbin has found a very closely related issue: #3060

**Comment by poikilotherm** [comment]
While writing the HERMES concept paper, we once more stumbled over this. I learned that Zenodo is offering such a schema at https://zenodo.org/schemas/deposits/records/legacyrecord.json

Recent talk with @atrisovic also revealed this would be very nice to create a crosswalk CodeMeta <-> Dataverse JSON

**Comment by djbrooke** [comment]
Thanks @poikilotherm - this has come up as we've discussed integration with some Harvard library systems as well. PRs welcome if you have the availability to work on this.

**Comment by skasberger** [comment]
This is a central part for a future release, as mapping between different schemas and data types seem to be of high relevance for the future.  You can find out more about my thoughts and work on this in pyDataverse here: https://github.com/gdcc/pyDataverse/issues/102

**Comment by pdurbin** [comment]
Related:

- #3621

**Comment by poikilotherm** [comment]
I just had a lovely chat with @JR-1991 while he was still in Boston about this (and other things).

For [easyDataverse](https://github.com/gdcc/easyDataverse), he would welcome having 

1. one general JSON Schema that describes and validates the general parts of a Dataset JSON and it's overall structure but not looking at the semantics of the metadata and
2. downloadable JSON Schema per collection that also contain in-depth validators for the metadata fields (which is dependent on the collections configuration).

The general schema might need to be versioned, so downloadable via the instance, too.

NOTE: It might help adding the latest version to https://www.schemastore.org to have autocomplete when writing JSON manually.

**Comment by qqmyers** [comment]
FWIW: The current json has a lot of schema-ish info in it already - would it make sense to remove that and make it flatter like the json-ld format? More work up front perhaps (and a breaking change) but a simpler schema and more readability.

**Comment by poikilotherm** [comment]
Yes, absolutely! We talked about that, too, but I forgot. We think alike you - meta information about the docs structure should be in a schema, not in a dataset's JSON.

**Comment by qqmyers** [comment]
This could potentially sync the processing of the json and json-ld - e.g. if the json-ld can be made to look just like the json with an added @context which I think could be possible.

**Comment by beepsoft** [comment]
I don't know if it helps, but you may also consider CEDAR as an example. Their metadata schema (https://more.metadatacenter.org/tools-training/outreach/cedar-template-model) is described as a JSON Schema, which defines the structure of a JSON-LD document, which should be the actual filled metadata.

**Comment by pdurbin** [comment]
@beepsoft you're actually using CEDAR with Dataverse, right? I forget, did you give us a sample we can play with? 😄

**Comment by beepsoft** [comment]
@pdurbin Yes I am. :-) As part of this we now have or will have soon the following tools:

- CEDAR template JSON Schema import as Dataverse metadata block
- Dataverse mdb export back to TSV
- Dataverse mdb export to CEDAR template JSON Schema (via the TSV format)

We use these to keep DV metadata blocks and CEDAR templates in sync.

I'm not sure if I gave you an example CEDAR template, but here is one, for example:

https://openview.metadatacenter.org/templates/https:%2F%2Frepo.metadatacenter.org%2Ftemplates%2Fdc3fa214-88f4-49dd-b56b-f4552b2d3474

The the end of the page you can find "Advanced View", there you can copy the JSON Schema.

For more examples,  you can easily register at

https://cedar.metadatacenter.org/

with github or ORCID account, and then can take a look at the public templates:

https://cedar.metadatacenter.org/dashboard?sharing=shared-with-everybody&folderId=https:%2F%2Frepo.metadatacenter.org%2Ffolders%2F6a419e8c-7f49-4426-9c8d-320de6ec061f

This is my usual test template:

CEDAR-NCBI Human Tissue

https://cedar.metadatacenter.org/templates/edit/https://repo.metadatacenter.org/templates/dc3fa214-88f4-49dd-b56b-f4552b2d3474?sharing=shared-with-everybody&folderId=https:%2F%2Frepo.metadatacenter.org%2Ffolders%2F6a419e8c-7f49-4426-9c8d-320de6ec061f

Once you go to the editor, at the bottom you can always see the JSON Schema associated with the template.

**Comment by poikilotherm** [comment]
>     * We need @poikilotherm in on a discussion and sizing.
:wink:

>       * We do already have a JSON schema library that knows how to validate a schema. Used in prov.json
A comment:  Yes, we have. It's old and outdated and should be replaced. Can be done in the same go.

I'm 65% sure that library is also not capable of *creating* a schema, just do validation of JSON with a given schema. So we might need to write out the schema using some JSON-P.

**Comment by pdurbin** [comment]
@beepsoft thanks! Interesting!

**Comment by mreekie** [comment]
Sizing:
- This is going to need to be a parent issue with smaller steps

**Steps:**
- Agree on general structure - create a schema that matches our dataset JSON without adding too much specifics on the metadatda. Check the Structure. (no code) (10)
- Use this schema to update the code.  We currently have a library that checks the JSON schema (wc3provenance schema). As part of this solution we should not feel limited to using our current library. (On the page about the JSON schema there are multiple java library options available for validation.)(10)
- Create API endpoints to retrieve a JSON schema that also has info about the required meta-data so that we can create a full-fledged validation of a dataset. That API end point has a paramenter that is the dataverse collection that you are targetting for deposit. (10)

reference:
- http://json-schema.org/implementations.html#validators

**Next Steps:**
- Set this issue up as a deliverable (mike)
- Create the 3 issues. ( @poikilotherm )

**Comment by mreekie** [comment]
> **Next Steps:**
> 
>  Set this issue up as a deliverable (mike)
> * Create the 3 issues. ( @poikilotherm )

@poikilotherm I've set this up as a backlog deliverable.

When you create the three issues:
- [ ] Add the label: "D: ImproveJsonValidation" to each
- [ ] Add the label: Size: 10 to each
- [ ] Add them directly to [the backlog](https://github.com/orgs/IQSS/projects/34/views/1)  to the column "SPRINT READY"

At that point they will be queued for an upcoming sprint.

Thank you so much stepping up during the meeting to create the issues!
